src/ServicePulse.Host/vue/src/components/DataView.vue

```diff
@@ -0,0 +1,173 @@
+import { computed, onMounted, onUnmounted, ref } from "vue";
+import { useFetchFromServiceControl } from "../composables/serviceServiceControlUrls";
+
+const props = defineProps({
+  apiUrl: String,
+  itemsPerPageOptions: {
+    type: Array,
+    default: () => [20, 35, 50, 75],
+  },
+  itemsPerPage: {
+    type: Number,
+    default: 50,
+  },
+  autoRefresh: {
+    type: Number,
+    default: 0,
+  },
+  showPagination: {
+    type: Boolean,
+    default: true,
+  },
+  showItemsPerPage: {
+    type: Boolean,
+    default: false,
+  },
+});
+
+const refreshTimer = ref();
+const items = ref([]);
+const pageNumber = ref(1);
+const itemsPerPage = ref(props.itemsPerPage);
+const totalCount = ref(0);
+const showPagination = computed(() => {
+  return props.showPagination && numberOfPages.value > 1;
+});
+
+const numberOfPages = computed(() => {
+  return Math.ceil(totalCount.value / itemsPerPage.value);
+});
+
+const showItemsPerPage = computed(() => {
+  return props.showItemsPerPage;
+});
+
+function changeItemsPerPage(value) {
+  itemsPerPage.value = value;
+  loadData();
+}
+
+function calculatePageNumbers() {
+  if (numberOfPages.value <= 10) {
+    return Array.from(Array(numberOfPages.value), (_, index) => index + 1);
+  }
+
+  let startIndex = pageNumber.value - 5;
+  let endIndex = pageNumber.value + 5;
+  if (startIndex < 0) {
+    // Increase the end index by the offset
+    endIndex -= startIndex;
+    startIndex = 0;
+  }
+
+  if (endIndex > numberOfPages.value) {
+    endIndex = numberOfPages.value;
+  }
+
+  return Array.from(Array(endIndex - startIndex), (_, index) => index + startIndex + 1);
+}
+
+function nextPage() {
+  pageNumber.value = pageNumber.value + 1;
+  if (pageNumber.value > numberOfPages.value) {
+    pageNumber.value = numberOfPages.value;
+  }
+  loadData();
+}
+
+function previousPage() {
+  pageNumber.value = pageNumber.value - 1;
+  if (pageNumber.value == 0) {
+    pageNumber.value = 1;
+  }
+  loadData();
+}
+
+function setPage(page) {
+  pageNumber.value = page;
+  loadData();
+}
+
+async function loadData() {
+  try {
+    const response = await useFetchFromServiceControl(`${props.apiUrl}?page=${pageNumber.value}&per_page=${itemsPerPage.value}`);
+    if (response.ok) {
+      totalCount.value = parseInt(response.headers.get("Total-Count"));
+      const data = await response.json();
+      items.value = data;
+    }
+  } catch (error) {
+    console.log(error);
+  }
+}
+
+function startRefreshTimer() {
+  if (props.autoRefresh > 0) {
+    refreshTimer.value = setInterval(() => {
+      loadData();
+    }, props.autoRefresh);
+  }
+}
+
+function stopRefreshTimer() {
+  if (refreshTimer.value) {
+    clearInterval(refreshTimer.value);
+  }
+}
+
+onMounted(() => {
+  startRefreshTimer();
+  loadData();
+});
+
+onUnmounted(() => {
+  stopRefreshTimer();
+});
+</script>
+
+<template>
+  <slot name="data" v-bind="items"></slot>
+  <div class="row">
+    <div v-if="showItemsPerPage" class="pagination col-md-2">
```

Comment:
is this div going to be extracted to a component?

Reply:
Extracted